File: 백준/Silver/14889. 스타트와 링크/스타트와 링크.java
import java.io.*;
import java.util.*;

public class Main {
    static int[] sumRow;
    static int[] sumCol;
    static boolean[] visit;
    static int N;
    static int min=Integer.MAX_VALUE;
    public static void main(String[] args) throws IOException {
        BufferedReader br=new BufferedReader(new InputStreamReader(System.in));
        N=Integer.parseInt(br.readLine());
        int sum=0;
        sumRow=new int[N];
        sumCol=new int[N];
        visit=new boolean[N];

        StringTokenizer st;
        for (int i=0; i<N; i++){
            st=new StringTokenizer(br.readLine());
            for (int j=0; j<N; j++){
                int status=Integer.parseInt(st.nextToken());
                sum+=status;
                sumRow[j]+=status;
                sumCol[i]+=status;
            }
        }

        dfs(0, 0, sum);
        System.out.println(min);
        
        br.close();
    }
    static void dfs(int pos, int depth, int diff){
        if (depth==N/2){
            min=Math.min(min, Math.abs(diff));
            return;
        }

        for (int i=pos; i<N; i++){
            if (!visit[i]){
                visit[i]=true;
                dfs(i+1, depth+1, diff-sumCol[i]-sumRow[i]);
                visit[i]=false;
            }
        }
    }
}

File: C++17/백준/Silver/14889. 스타트와 링크/스타트와 링크.cc
#include <iostream>
#include <cmath>
#include <climits>
using namespace std;

int N, min_res = INT_MAX;
int R[21];
int C[21];
bool visited[21];

void dfs(int pos, int depth, int val){
    if (depth == N / 2){
        min_res = min(min_res, abs(val));
        return;
    }

    for (int i = pos; i < N; i++){
        if (!visited[i]){
            visited[i] = 1;
            dfs(i + 1, depth + 1, val - R[i] - C[i]);
            visited[i] = 0;
        }
    }
}

int main(){
    cin >> N;

    int sum = 0;
    for (int i = 0; i < N; i++){
        for (int j = 0; j < N; j++){
            int x; cin >> x;
            sum += x;
            R[i] += x;
            C[j] += x;
        }
    }

    dfs(0, 0, sum);
    cout << min_res << '\n';
    
    return 0;
}